---
title: "Gene.Classifier"
format: html
toc: true
---



## Import Data

```{r}
library(dplyr)
library(magrittr)

raw_data_dir <- "C:/Users/User/Documents/UCSF/BREATH2/pediatric-mNGS-LRTI-classifier/data/raw"
data_dir <- "C:/Users/User/Documents/UCSF/BREATH2.Prot/data"
results_dir <- "C:/Users/User/Documents/UCSF/BREATH2.Prot/results"


metadata_df_ALL <- read.csv(file.path(raw_data_dir, "sample_metadata.csv"), stringsAsFactors=F)

JCI_expanded_metadata <- read.csv(file.path(data_dir, "fixed_metadata_10242025.csv"))

vap_soma_normalized <- read.csv(file.path(data_dir, "vap_soma_normalized.csv"))





#------------------------------------------------------------------------------#




# Include only the 150 patients with both transcriptomic and proteomic data
patients_vector <- intersect(vap_soma_normalized$patient_id, 
                             JCI_expanded_metadata$SubjectID)

Prot.Data <- vap_soma_normalized[vap_soma_normalized$patient_id %in% patients_vector, ]
Meta.Data <- JCI_expanded_metadata[JCI_expanded_metadata$SubjectID %in% patients_vector, ]


# Ensures rows of Meta.Data match those of Prot.Data
Meta.Data <- Meta.Data[match(Prot.Data$patient_id, Meta.Data$SubjectID), ]

# Produces data frame with patients, age_in_years, sex, and LRTI_Adjudication
patient_info_ALL <- data.frame(
  sample_name = Meta.Data$sanitized_name,
  patient_id = Prot.Data$patient_id,
  sex = Meta.Data$sex,
  age_in_years = Meta.Data$age_in_years,
  LRTI_adjudication = Meta.Data$LRTI_adjudication,
  stringsAsFactors = FALSE
)

# Restrict metadata_df to same 151 patients used in patient_info_ALL
metadata_df <- metadata_df_ALL %>%
  filter(sample_name %in% patient_info_ALL$sample_name)

nrow(metadata_df)  # should be 151
all(metadata_df$sample_name == patient_info_ALL$sample_name)  # should be TRUE

# Write csv
write.csv(metadata_df, file.path(data_dir, "metadata_df.csv"), row.names = FALSE)



#------------------------------------------------------------------------------#
```



# Fold Assignment


```{r}
library(dplyr)
library(magrittr)

metadata_df <- read.csv(file.path(data_dir, "metadata_df.csv"), stringsAsFactors=F)


# Restrict to Definite/No Evidence samples for generating cross-validation folds
metadata_df %>%
  dplyr::filter(LRTI_adjudication %in% c("Definite", "No Evidence")) ->
  cv_samples

# Generate folds for 5-fold cross-validation; ensure each fold has >=9 No Evidence samples
success <- FALSE
while (!success) {

  cv_samples %>%
    dplyr::select(sample_name, LRTI_adjudication) %>%
    dplyr::mutate(fold=sample(rep(1:5, length.out=nrow(.)))) ->
    cv_folds

  cv_folds %>%
    dplyr::select(LRTI_adjudication, fold) %>%
    table() ->
    cv_fold_table

  print("Generated CV folds with following counts:")
  print(cv_fold_table)

  cv_fold_table %>%
    .["No Evidence", ] %>%
    min() %>%
    {. >= 9} ->
    success

  if (!success) {
    print("At least one fold has too few No Evidence samples. Regenerating folds...")
  }
}

# Add folds to sample metadata and output the table
metadata_df %>% 
  dplyr::left_join(cv_folds %>% select(sample_name, fold), by = "sample_name") %>% 
  write.csv(classifier_meta_csv, row.names=F)


  
```

